### PR TITLE
feat: Improved array arithmetic support

### PR DESCRIPTION
**Pull request by nameexhaustion**
* Adds support for performing arithmetic between numeric and array columns. The numeric value is broadcasted horizontally to every array element within each row, similar to list arithmetic.
* Broadcasting for arithmetic between 2 array columns is now done without materializing the input columns
* Adds support flooring division for array columns
* Fixes an issue with incorrect outer validities when adding arrays (https://github.com/pola-rs/polars/issues/19838)

```python
shape: (6, 8)
┌────────────────────┬──────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬──────┬────────────────────┐
│ a                  ┆ p    ┆ a * p              ┆ a * first(p)       ┆ first(a) * p       ┆ a // p             ┆ q    ┆ a + q              │
│ ---                ┆ ---  ┆ ---                ┆ ---                ┆ ---                ┆ ---                ┆ ---  ┆ ---                │
│ array[i64, (1, 2)] ┆ i64  ┆ array[i64, (1, 2)] ┆ array[i64, (1, 2)] ┆ array[i64, (1, 2)] ┆ array[i64, (1, 2)] ┆ i64  ┆ array[i64, (1, 2)] │
╞════════════════════╪══════╪════════════════════╪════════════════════╪════════════════════╪════════════════════╪══════╪════════════════════╡
│ [[1, 1]]           ┆ 1    ┆ [[1, 1]]           ┆ [[1, 1]]           ┆ [[1, 1]]           ┆ [[1, 1]]           ┆ null ┆ [[null, null]]     │
│ [[2, 2]]           ┆ null ┆ [[null, null]]     ┆ [[2, 2]]           ┆ [[null, null]]     ┆ [[null, null]]     ┆ null ┆ [[null, null]]     │
│ [[3, 3]]           ┆ 3    ┆ [[9, 9]]           ┆ [[3, 3]]           ┆ [[3, 3]]           ┆ [[1, 1]]           ┆ null ┆ [[null, null]]     │
│ [[null, null]]     ┆ 4    ┆ [[null, null]]     ┆ [[null, null]]     ┆ [[4, 4]]           ┆ [[null, null]]     ┆ null ┆ [[null, null]]     │
│ [null]             ┆ 5    ┆ [null]             ┆ [null]             ┆ [[5, 5]]           ┆ [null]             ┆ null ┆ [null]             │
│ null               ┆ 6    ┆ null               ┆ null               ┆ [[6, 6]]           ┆ null               ┆ null ┆ null               │
└────────────────────┴──────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴──────┴────────────────────┘
```
<details>
<summary>Table code</summary>

```python
df = (
    pl.select(
        pl.lit(
            pl.Series(
                [[[1, 1]], [[2, 2]], [[3, 3]], [[None, None]], [None], None],
                dtype=pl.Array(pl.Array(pl.Int64, 2), 1),
            )
        ).alias("a"),
    )
    .with_columns(
        pl.int_range(1, 1 + pl.len())
        .pipe(lambda e: pl.when(e != 2).then(e))
        .alias("p"),
    )
    .with_columns(
        (pl.col("a") * pl.col("p")).alias("a * p"),
        (pl.col("a") * pl.first("p")).alias("a * first(p)"),
        (pl.first("a") * pl.col("p")).alias("first(a) * p"),
        (pl.col("a") // pl.col("p")).alias("a // p"),
    )
    .with_columns(
        pl.lit(None, dtype=pl.Int64).alias("q"),
    )
    .with_columns(
        (pl.col("a") + pl.col("q")).alias("a + q"),
    )
)

print(df)
```
</details>

Fixes https://github.com/pola-rs/polars/issues/19356
Fixes https://github.com/pola-rs/polars/issues/19838

### CODE REVIEW

File: crates/polars-core/src/series/arithmetic/fixed_size_list.rs [hunk 1]
```diff
@@ -0,0 +1,820 @@
+use polars_error::{feature_gated, PolarsResult};
```
Comment (nameexhaustion) [inline crates/polars-core/src/series/arithmetic/fixed_size_list.rs:1]: Note: This file has mostly the same structure as the one for list arithmetic.

File: crates/polars-core/src/series/arithmetic/fixed_size_list.rs [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+            //                  |  L  |  N  |  R  | // Broadcast (L)eft, (N)oBroadcast, (R)ight
+            // ListToList       | [1] |  0  |  1  |
+            // ListToPrimitive  |  2  |  3  |  4  |
+            // PrimitiveToList  | [4] | [3] | [2] |
+
+            self.swapped = true;
+
+            match (&self.op_apply_type, &self.broadcast) {
+                // Mostly the same as ListNumericOp, however with fixed size list we also have
+                // (BinaryOpApplyType::ListToPrimitive, Broadcast::Left) as a physical impl.
```
Comment (nameexhaustion) [inline crates/polars-core/src/series/arithmetic/fixed_size_list.rs:349]: Arrays have a simpler layout compared to lists, so we can also implement `array<->primitive` with the array side being broadcasted without materializing.

File: crates/polars-core/src/series/arithmetic/list_borrowed.rs [hunk 3]
```diff
@@ -3,38 +3,58 @@
...
     }
 
     fn remainder(lhs: &ListChunked, rhs: &Series) -> PolarsResult<Series> {
-        NumericListOp::Rem.execute(&lhs.clone().into_series(), rhs)
+        NumericListOp::rem().execute(&lhs.clone().into_series(), rhs)
     }
 }
 
-#[derive(Debug, Clone)]
-pub enum NumericListOp {
```
Comment (nameexhaustion) [inline crates/polars-core/src/series/arithmetic/list_borrowed.rs:31]: Factored out this and some other parts of list arithmetic for re-use

File: crates/polars-core/src/series/arithmetic/list_borrowed.rs [hunk 4]
```diff
@@ -455,59 +310,56 @@ mod inner {
                     self.swapped = false;
                     self._finish_impl_dispatch()
                 },
-                (BinaryOpApplyType::PrimitiveToList, Broadcast::Right) => {
-                    // We materialize the list columns with `new_from_index`, as otherwise we'd have to
-                    // implement logic that broadcasts the offsets and validities across multiple levels
-                    // of nesting. But we will re-use the materialized memory to store the result.
+                (BinaryOpApplyType::ListToList, Broadcast::Left) => {
```
Comment (nameexhaustion) [inline crates/polars-core/src/series/arithmetic/list_borrowed.rs:313]: drive-by - changed the order of this enum matching to the same order of the comment (expand the diff above to see the comment)

File: py-polars/tests/unit/datatypes/test_array.py [hunk 5]
```diff
@@ -383,16 +383,3 @@ def test_zero_width_array(fn: str) -> None:
 
                 df = pl.concat([a.to_frame(), b.to_frame()], how="horizontal")
                 df.select(c=expr_f(pl.col.a, pl.col.b))
-
-
-def test_elementwise_arithmetic_19682() -> None:
```
Comment (nameexhaustion) [inline py-polars/tests/unit/datatypes/test_array.py:388]: moved

File: py-polars/tests/unit/operations/arithmetic/test_array.py [hunk 6]
```diff
@@ -1,7 +1,290 @@
...
+@pytest.mark.parametrize(
+    "array_side", ["left", "left3", "both", "both3", "right3", "right", "none"]
+)
+@pytest.mark.parametrize(
+    "broadcast_series",
+    BROADCAST_SERIES_COMBINATIONS,
+)
+@pytest.mark.parametrize("exec_op", EXEC_OP_COMBINATIONS)
+@pytest.mark.slow
+def test_array_arithmetic_values(
```
Comment (nameexhaustion) [inline py-polars/tests/unit/operations/arithmetic/test_array.py:25]: Same parametrized test copied from list arithmetic

**Review by ritchie46 — APPROVED** [review]
